### PR TITLE
pull

### PR DESCRIPTION
**Pull request by brock7**
This change is [Reviewable](https://reviewable.io/reviews/x64dbg/x64dbg/1619)

### CODE REVIEW

**Comment by mrexodia** [comment]
@blaquee 

> <torusrxxx> in pull request #1537 he wants to use a custom and more powerful OpenProcess
> <mrexodia> he can just hook TitanOpenProcess

See https://botbot.me/freenode/x64dbg/2017-04-13/?msg=83944743&page=1

File: src/dbg/debugger_commands.cpp [hunk 1]
```diff
@@ -1199,7 +1199,7 @@ CMDRESULT cbDebugAttach(int argc, char* argv[])
     if(!GetModuleFileNameExW(hProcess, 0, wszFileName, MAX_PATH))
     {
         dprintf("Could not get module filename %X!\n", pid);
-        return STATUS_ERROR;
+        // return STATUS_ERROR;
```
Comment (mrexodia) [inline src/dbg/debugger_commands.cpp]: Could you give more details about this bug?

File: src/dbg/module.cpp [hunk 2]
```diff
@@ -82,6 +82,9 @@ bool ModLoad(duint Base, duint Size, const char* FullPath)
             strcpy_s(file, FullPath);
     }
 
+	if (file[0] == '\0')
```
Comment (mrexodia) [inline src/dbg/module.cpp]: If this to not load modules that don't have a name?

File: src/dbg/stringutils.cpp [hunk 3]
```diff
@@ -24,6 +24,9 @@ StringList StringUtils::Split(const String & s, char delim)
 
 String StringUtils::Escape(const String & s)
 {
+	// FIXME: bug
```
Comment (mrexodia) [inline src/dbg/stringutils.cpp]: Could you explain the bug?

File: src/dbg/thread.cpp [hunk 4]
```diff
@@ -241,8 +242,10 @@ int ThreadSuspendAll()
     int count = 0;
     for(auto & entry : threadList)
     {
-        if(SuspendThread(entry.second.Handle) != -1)
-            count++;
+		if (SuspendThread(entry.second.Handle) != -1)
+			count++;
+		else
+			dprintf("thread(%d) cannot suspend. handle: %x", entry.second.ThreadId, entry.second.Handle);
```
Comment (mrexodia) [inline src/dbg/thread.cpp]: This has been added into x64dbg:

```c++
int ThreadSuspendAll()
{
    // SuspendThread does not modify any internal variables
    SHARED_ACQUIRE(LockThreads);

    int count = 0;
    for(auto & entry : threadList)
    {
        if(SuspendThread(entry.second.Handle) != -1)
            count++;
        else
            dprintf(QT_TRANSLATE_NOOP("DBG", "Failed to suspend thread 0x%X...\n"), entry.second.ThreadId);
    }

    return count;
}
```

File: x64dbg_note.txt [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+		tokenPrivileges.Privileges[0].Attributes = 0;
+	AdjustTokenPrivileges(hToken, FALSE, &tokenPrivileges, sizeof(TOKEN_PRIVILEGES), NULL, NULL);
+	dwLastError = GetLastError();
+	CloseHandle(hToken);
+	return dwLastError;
+}
+
+HANDLE _OpenProcess(DWORD dwDesiredAccess, bool bInheritHandle, DWORD dwProcessId)
+{
+	HANDLE hCsrss = OpenProcess(PROCESS_DUP_HANDLE, FALSE, 428);
```
Comment (mrexodia) [inline x64dbg_note.txt]: Could you explain this hardcoded 428 process id?

File: x64dbg_note.txt [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+		dprintf("DuplicateHandle() failed\n");
+		return NULL;
+	}
+
+	dprintf("OpenProcess() ok. handle = %x\n", h2);
+	return h2;
+	
+}
+
+HANDLE MyOpenProcess(DWORD dwDesiredAccess, bool bInheritHandle, DWORD dwProcessId)
```
Comment (mrexodia) [inline x64dbg_note.txt]: I think it should be possible to write a plugin that hooks `TitanOpenProcess` and calls this instead...

**Comment by Mattiwatti** [comment]
Re: the idea of using CSRSS as a trampoline to duplicate kernel mode protected DRM process handles mentioned on IRC: that won't work. The same mechanism that is responsible for notifying kernel mode drivers of `NtOpenProcess` calls ([ObRegisterCallbacks](https://msdn.microsoft.com/en-us/library/windows/hardware/ff558692(v=vs.85).aspx)) can be used in the same way for `NtDuplicateObject` (= `DuplicateHandle`), or in fact any system call involving a handle that the driver wants to monitor. That is apart from the fact that opening a useable handle to CSRSS itself is no longer possible since Windows 8.1, even with debug privileges enabled like in this PR. I wrote [PPLKiller](https://github.com/Mattiwatti/PPLKiller) because of this. It should already be linked in the wiki, but it's not exactly a solution for everyone since it runs in kernel mode. Generally speaking, trying to circumvent any sort of kernel mode restriction from user mode is going to be a very big waste of time.

**Comment by brock7** [comment]
pull

**Comment by mrexodia** [comment]
There are no changes relevant to the current build in this pull request and there are also conflicts. I cannot merge this in any way.

**Pull request closed without merging**